### ISSUE TITLE
2.4.0 development process

### ISSUE THREAD

**Issue by sthalik** [opentrack/opentrack#688]
`last modified on 20180208`

Hey,

I've since started work on 2.4.0. Let's outline some of the new features:

In the `20171013` build:

- [x] 🍺  PT tracker: different modes for grayscaling
- [x] 🔧 Octopus: no longer has artifacts at the bottom or elsewhere
- [x] 🍺 PT tracker: start amcap for exposure control for the PS3 Eye.
- [x] 🍺 Full portable mode setup (see wiki)
- [x] 🍺 PT current threshold value description

In the `20171031` build:

- [x] 🍺 Use buffered joystick and keyboard press events. The listener now runs 10 times a second and not 250 times a second (Windows).
- [x] 🔧 Fix portable mode global settings saving
- [x] 🔧 Fix bugs in X-Plane plugin for Linux and OSX (buffer overflow, etc.)
- [x] 🍺 Annoy users who run as `root` on Linux
- [x] 🍺 Allow setting max output pitch value to 90 or 180. Existing configs are migrated based on their existing values.
- [x] 🍺 Accela allows for double the deadzone value
- [x] 🍺 Accela rotation nonlinearity was euthanized
- [x] 🔧 Fix mapping background color mismatch

In the `20180104` build:

- [x] :beer: :wrench: Fix exposure dialog for PS3 Eye.
- [x] 🔧 Notice of obsolescence for Windows XP support since now.
- [x] 🔧 Notice of obsolence for CPUs without SSE3 support from now on. All Intel and AMD chipsets since 2004 support SSE3.
- [x] 🔧 Allow all module types to output status and an optional error message. See `bool IProtocol::correct()`. Make it possible to send error descriptions and have the `runtime_libraries` class manage the error dialog stuff.

In the `20180110` build:

- [x] :beer: tracker/pt: reset dynamic pose on center
- [x] :beer: relative translation will work only when looking backward or downward (#712)
- [x] :wrench: fix freetrack deactivation method
- [x] :octopus:  [feed Octopus](https://www.youtube.com/watch?v=8h-EdX0_vmk)

In the `20180116` build:

- [x] 🍺 support HiDPI screens without blurry fonts
- [x] 🍺 mappings: add more guides depending on max X/Y on the graph
- [x] 🔧 minor freetrack fix

In the `20180119` build:

- [x] 🍺 add a new mode for relative translation, tries to detect aiming precisely, disabling relative translation for the time. "aiming precisely" is defined as looking straight toward the center.
- [x] 🔧🐈 thanks to @cpuwolf for noticing PT has no images.
- [x] 🔧 heavy internal work in prepation for the Wiimote tracker, reusing parts of the PT tracker.

In the `20180205` build:

- [x] 🍺🐈 add Wiimote tracker, by @cpuwolf
- [x] 🔧 fix PT not showing camera options when tracking stopped

In the `20180216` build:

- [x] 🔧 Change "enabled when not aiming" reltrans mode limits. Now is up to  50 yaw, 40 pitch.
- [x] ~~🍺 Accela: experimental minor movement when within deadzone~~.
- [x] 🔧 🐈 Thanks to Artem Sobolev @ljgdasfhk for spotting build errors on Linux.
- [x] 🔧 HiDPI fix with varying display resolutions.
- [x] 🍺 🐈 Wiimote tracker automatic pairing! Work done by @cpuwolf.
- [x] 🔧 Invalid hatire input outside normal (-180)->180 bounds workaround.
- [x] 🍺 Euthanize Accela smoothing. Rename "sensitivity" to "smoothing".

Next build:

- [x] 🔧 Close button must exit even with tray enabled. Restore that behavior (reported by @albertolg).

- - -

Blocking issues:

- [ ] 🔧 Test the new `NPClient.dll` fully and send it to anti-cheat program vendors. It needs testing the timer code and overall logic where the timer is used, as well as return values from `GetData`.
- [ ] 🍺 Support for the Tobii EyeX devices. For now, let's only do accumulative mode with non-configurable spline. From 2 to 7 yaw/pich degrees per second max, default to 3. Limit tracker output yaw and pitch to 45, 45 for new users.
- [ ] 🔧 Greatly simplified the pipeline code. Needs testing all pipeline features fully.

Not done, not blocking:

- [ ] 🍺 Support for arbitrary plugins at various stages (**STARTED**)
- [ ] 🍺 Support for joystick input on Linux and OSX via libevdev

**Comment by albertolg** [comment]
is too hard to implement a 3d head like was in freetrack instead of the octopus and disable it when freetrack is minimized to save the cpu ?

**Comment by sthalik** [comment]
It's time-consuming. I personally have no interest in this. It's an improvement however, and if someone sends the code, I will accept it.

**Comment by Len62** [comment]
So with amcap implemented we would no longer need the cleye driver, as far as OT is concerned ?

**Comment by sthalik** [comment]
It's not a driver. It's an app that shows camera options successfully.

**Comment by Len62** [comment]
Got it, thanks!

**Comment by sthalik** [comment]
Replacing the CL-Eye driver is by all means possible -- Linux has the right byte sequences to emit in its own driver. Someone could reuse those in a WinUSB driver. But it's a lot more work than finally-working filter options.

**Comment by Len62** [comment]
Having working filter options will be great. I generally do not have to change mine but for others it will be nice to be able to adjust options in OT.

**Comment by sthalik** [comment]
I found something strange. See <<http://www.driverscape.com/download/usb-camera-b4.09.24.1>>. The **BOTTOM** link. Top is some adware. Is that a free ps3 eye driver?

**Comment by Len62** [comment]
The cleye driver is version 5.3.0.341, device ID# reported for windows is usb-camera-b4.09.24.1. Looks like an earlier driver version to me.

**Comment by sthalik** [comment]
Yeah, but free and aparently legal.

**Comment by Len62** [comment]
Free is always good! I may give it a try just to see if there are any big differences.

**Comment by albertolg** [comment]
@sthalik I found this https://github.com/HipsterSloth/psmove-unity5/wiki/PSEye-Software-Setup-%28Windows%29

the camera is working, but opentrack don't see it

**Comment by guillerodriguezv** [comment]
Hi, excellent work,  may I ask for Pimax support in your next release,  please. 

Thanks

**Comment by albertolg** [comment]
@sthalik it's possible to use the libusb drivers with the ps3eye so we can get rid of the old codelabs drivers?

**Comment by sthalik** [comment]
@albertolg yes, and I'm looking at libusbK specifically. The API isn't braindead unlike Microsoft's. It can be used inside a custom DirectShow filter for the webcam.

**Comment by albertolg** [comment]
great news @sthalik thanks

**Comment by sthalik** [comment]
I'm gonna add it and disable by default.

Also, I added "portable installation" mode. Creating "portable.txt" in the software's directory turns it on. Configs are gonna get created in the program's directory.

**Comment by Len62** [comment]
Sounds good. Auto update to only release versions, correct?

**Comment by sthalik** [comment]
Yes of course.

**Comment by sthalik** [comment]
You should like this:

![image](https://user-images.githubusercontent.com/1896811/31552683-7d1ab96e-b038-11e7-8ec2-cddc73dd5849.png)
![image](https://user-images.githubusercontent.com/1896811/31552744-9abaecfa-b038-11e7-80da-8684064e9ebc.png)

**Comment by petsild** [comment]
Yeah awesome

**Comment by Len62** [comment]
Very nice!

**Comment by 503Kento** [comment]
Agreed, very nice, should help with my uneven LED brightness that I'm too lazy to fix.

**Comment by sthalik** [comment]
Grab it as usual from <<http://opentrack.misaki.pl>>.

**Comment by Len62** [comment]
Test build performs really great. It was nice to be able to use amcap to get to the ps3 eye camera settings, and the color channel options are a great addition. The changes to the point extraction settings will make it easier for users to find that "sweet-spot".

**Comment by sthalik** [comment]
@Len62 I intend to make the `amcap` dialog automatically use the right camera and its settings, pop up the filter options and exit when they're closed. It's just an old app and very annoying to deal with. Simply disabling options like MPEG2 capture, sound, saved settings, etc. took way more than it should.

Also we made a Discord channel with @mrsanchos. You can drop by to discuss development if you want. The link -- <<https://discord.gg/dNS2hhB>>.

**Comment by MathijsG** [comment]
Somehow amcap isn't opening here with the test build. Can I debug something for you?

**Comment by Len62** [comment]
Doesn't open at all, or screen is blank? If OT is running then the amcap window opens w/blank screen. If you stop OT then amcap opens correctly, at least for me, on the newest test build.

**Comment by MathijsG** [comment]
Nothing opens at all. That's the weird thing. While the separated executable in the same test folder of latest Opentrack test, does open.

**Comment by Len62** [comment]
Weird for sure. @sthalik will probably have an idea.

**Comment by sthalik** [comment]
@MathijsG,

These aren't solutions but possible means of identifying the problem

- did antivirus pop up? Avast is known to whine about webcam usage
- check Event Viewer

**Comment by sthalik** [comment]
Can you check if my blog post is understandable <<https://sthalik.blogspot.com/2017/10/windows-timer-resolution-nuclear-option.html>> for non-programmers? Is it too verbose and without context? Thanks!

**Comment by sthalik** [comment]
I posted `opentrack-test-20171029_01-BACKUP_YOUR_CONFIGS.7z` to <<http://opentrack.misaki.pl>>.

**BACK UP YOUR CONFIGS**
**BACK UP YOUR CONFIGS**
**BACK UP YOUR CONFIGS**

**Comment by albertolg** [comment]
@sthalik any news about using libusbK instead of cl-eye-drivers? thanks

**Comment by sthalik** [comment]
@albertolg it's very hard compared to regular opentrack stuff. Don't wait on it.

**Comment by Len62** [comment]
Tried out the latest test build, backed-up my configs, and everything ran smoothly. The octopus seems a little more fluid, did not run into any problems with the config files. Even running the previous test version w/a config from the newest worked good. Will build a new config from scratch later on today for the new test version. Read your blog and I don't think non-programmers would find it too difficult to understand. I would probably get rid of the "F" word, makes it seem somewhat juvenile, I'm not offended by it, I spent 22 years in the military. Got a real kick out of seeing the assembler code, I had learned a little myself a long time ago, but then life got in the way and I gave it up. I wished I hadn't.

**Comment by sthalik** [comment]
The threshold slider settings aren't passed onto the old version. It only migrates old -> new. Just so you know.

Do you have any problems with shortcuts, either joystick or keyboard? Now it's ok to press a shortcut very quickly, it will register. But code changes break stuff as you've seen through the project's history.

> I would probably get rid of the "F" word, makes it seem somewhat juvenile

Thanks, you have a point. In Polish one can compose a sentence with 80% profanity and still transmit plenty of information. In English swearing looks lame, kinda.

**Comment by MathijsG** [comment]
@sthalik I understand your blog post, but I do see myself as having some knowledge of programming. However I miss the underlying problem. First you introduce the lack of resolution as a problem and then you jump to issues with Spotify. I'm assuming Spotify is using this 'fuck system performance' workaround you are mentioning? And you trying to undo that with your own DLL? Or am I totally wrong :-P.

**Comment by Len62** [comment]
I use two shortcuts - toggle = ctrl + stick button, center = shift + stick button, and have not noticed any issues. Sometimes w/OT running, and tracker off, if I start a game and then toggle the tracker on the game will lock-up. I figured the threshold settings were not passed back, but it did not have any adverse effect on the previous test version.

**Comment by MathijsG** [comment]
By the way is it possible to upload regular zip instead of 7-zip's own fileformat? Wanting to try this version on other laptop, but I end up having annoyances with Windows 10 apps for opening a .7zip file without having to bloat my laptop :-P.

**Comment by sthalik** [comment]
@Len62,

> Sometimes w/OT running, and tracker off, if I start a game and then toggle the tracker on the game will lock-up.

Looks like a logic error in the game's head tracking support. opentrack's code inside the game process is so simple it doesn't even treat having head tracking on or off separately. There's the matter of particular .csv game protocol version, but it's handled exactly once and in the same way, inside the game's process' opentrack library.

@MathijsG,

> By the way is it possible to upload regular zip instead of 7-zip's own fileformat?

Damn, 7-zip is the first thing I install, even before notepad++ :) You won't get anywhere without an archiver.

**Comment by Len62** [comment]
I did not think the hot-key issue was related to OT, more to the game. I just start the tracker before starting the game and then there are no issues w/the hot-keys. I thought about looking at Python just for the heck of it, don't know that I have the patience for programming nowadays.

**Comment by MathijsG** [comment]
Interesting: When I drop the test folder directly in C:\ it works perfectly, but when I drop it on my Desktop, which makes it E:\Users\Mathijs Groothuis\Desktop\ it doesn't open.

Maybe it's because of the spaces? Or maybe something that assumes everyone is working on C:\?

**Comment by Len62** [comment]
Just noticed that in the newest test version that the modifier keys are not being seen when setting hot-keys.

**Comment by sthalik** [comment]
@Len62 great find! I updated <<http://opentrack.misaki.pl>> and removed the build from two days ago. There are no other changes so it's the old build + the modifier fix.

Don't play Telltale's Game of Thrones! It's basically a humiliation simulator going on for 5 DLCs. No matter what you do in any story arcs, you lose. It should be a 5-minute game with the player being presented a game over screen right after the game starts. Especially don't pay for that shit.

**Comment by Len62** [comment]
Thanks! D/L'd the new test version but will have to wait to the morning to try it out. Have grandkids to take trick or treating tonight.  I only play two simulations, IL2-BOS and DCS World (mainly for the helos).

**Comment by sthalik** [comment]
Please check the `20171226` build on http://opentrack.misaki.pl.

#### Arma anticheat notes

For Arma anticheat use the .dll's in `arma-anticheat/` directory. Copy to `modules/` and overwrite. Otherwise PLEASE test the new dlls. They should free up mouse view when tracking is stopped. That's the purpose of the new .dll's.

Only after they're fully tested I'm sending for whitelisting.

I think we should go for `2.4.0` as the version number since so much has changed already.

### Important:

The freetrack npclient `.dll`'s need your heavy testing with each flight game you have. please test each game. the new feature is freeing up mouse control when opentrack is stopped/closed.

Please check for bugs in old functionality since internal changes could've broken things that worked.

I'm really tired and weak. I'm looking for a co-maintainer. I'd also not like the project to die because of my possible permanent inactivity. I'm looking for someone with a few years of hobby/work experience and some C++11 and up knowledge, or old-style C++ knowledge and willing to learn the new-style C++. We're using a subset of C++17 features.

In the past, people with decent experience like @DaMichel found their way around the codebase easily. You can too if you want.

### Changes:

- dialog with the reason shown in case a protocol/tracker/filter can't initialize. Not used for vjoystick or realsense cause they have better dialogs than a stock error window.
- heavy refactorings to make former possible
- other heavy refactorings to make codebase look better for programmers
- reorg settings dialog. comment if it got worse.
- speedup tracker/pt color to grayscale conversion
- default to natural grayscaling in tracker/pt since it's way fastest
- allow freetrack mouse control when disabled, see beginning of this message
- make spline snap value lower for translation mappings

**Comment by sthalik** [comment]
For testing freetrack with existing games, please:

- check with opentrack not started first
- stop opentrack, see if it resumes tracking after it's started again
- see if mouse control returns when starting again

Please list games you checked.

**Comment by MathijsG** [comment]
Will test tomorrow with the games I own. Hope you will get better soon mate. Sounds serious.  Unfortunately I'm not really a programmer, more of a UX and bug squatter.

**Comment by sthalik** [comment]
By the way, @MathijsG thank you for dozens of contributions you've made over the past years.

**Comment by Len62** [comment]
Will give this a test later today. Sorry to hear about your health issues, hoping you get better.

**Comment by MathijsG** [comment]
Euro Truck Simulator 2 not working. Will investigate further.

**Comment by Len62** [comment]
I only play IL2-BOS and DCS World, no issues with either one.

**Comment by MathijsG** [comment]
Somehow it didn't work with any program (FaceTrackNoIR, Trackhat, etc) so I restarted the game, then it worked perfectly. No issues with ETS2. Camera settings problem still exists.

**Comment by sthalik** [comment]
What camera settings issue?

**Comment by MathijsG** [comment]
See https://github.com/opentrack/opentrack/issues/688#issuecomment-339672961

The Camera Settings app doesn't open Amcap for my PS3 Eye camera. It just doesn't open anything.

When I put it in "C:\" everything does work. Also the amcap.exe executable does work seperately. Looks like something with spaces in the path maybe?

**Comment by sthalik** [comment]
@MathijsG I can verify it's a bug. There are missing quotation marks around the executable path.

But amcap needs a bit of work to select the right camera, resolution, fps, open the filter dialog and close once it's closed. I'm not done with this fully yet.

@MathijsG @Len62 have you checked whether your games track after you stop and restart tracking? Also it should return mouse control while it's disabled.

**Comment by sthalik** [comment]
I FIXED THE PS3 EYE CAMERA SETTINGS. IT WORKS FULLY WITHOUT AMCAP.

**Comment by Len62** [comment]
I did not have any issues stopping and restarting. I'm going to try the latest test build in a little while, and see how things go.

**Comment by MathijsG** [comment]
No issues yet with ETS2, will try more games when I am avoiding writing my thesis :-P. However the only 'issue' I encounter with ETS2 for example (and I remember more games in the past had this behavior) is that when you start Opentrack (I mean open the application) AFTER the game has been launched already you sometimes can't get the tracking to work in-game (the octopus all works and stuff) but this is also the case with the current version of Opentrack, Trackhat and FaceTrackNoIR, so maybe some games only load certain modules when they detect something TrackIR-like present on the system?

**Comment by Len62** [comment]
Tried the latest test build and it was nice to adjust the PS3 settings and view the changes in the OT output window! Nicely done! On another note, I've had some issues with this build shutting down while toggling it off and on in IL2-BOS. Also noticed in that game that returning to the center position left the in-game view off center. Thought maybe it was a setting in Bos, but the Dec test build, and the last release version do not show this, strange one. As far as mouse control is concerned, is it supposed to be disabled when running OT? I noticed that even with OT running the mouse still works moving the view around.

**Comment by sthalik** [comment]
@Len62 the new npclient.dll detects the "off" state as all 6 values being exact zeros. I guess there should be another way to designate an "off" state, otherwise the very last state before shutting down takes precedence.

We have some free data structure members so why not.

**Comment by sthalik** [comment]
@Len62, I changed the deactivation logic and it should now work.

**Comment by sthalik** [comment]
New build! See the first post for changes. As usual, binaries on <<http://opentrack.misaki.pl>>.

Only a few user-visible fixes but pretty important stuff.

**Comment by Len62** [comment]
Centering issue is fixed. However, still having issues with toggling off/on, and intermittent cases where OT is shutting down. If OT is running, but tracking off, I start IL2-BOS, toggle tracking on w/o any problems. However, if I toggle it off, and then back on, it takes anywhere from 5 to 15 seconds to respond, or it doesn't respond at all and OT completely shuts down. I'll do some more testing but so far the problem exists if I just use keyboard shortcuts, or modifier key + joystick button.

**Comment by sthalik** [comment]
@Len62 check the updated `04` build for yesterday. I changed the logic for centering and deactivating the tracker. Stopping tracking isn't instant for me with Falcon BMS but it's the game's fault, not opentrack's. I send the code for getting it disabled in just 2 frames.

There was a possible race where getting tracking stopped could keep it stopped randomly for more than a single frame, especially on high FPS or CPUs with few cores.

Are the freetrack issues new?

Tracking should now consistently reset back to zero pose when it's getting stopped.

**Comment by sthalik** [comment]
I meant to send it yesterday but fell asleep. Damn.

**Comment by Len62** [comment]
This new build is working fine. No problems with returning to zero when stopping, and no intermittent shut-downs of OT. I couldn't recreate the freetrack issue with this build. Sometimes I believe some issues are due to the games themselves. Thanks for the quick fix.

**Comment by MathijsG** [comment]
Did you include the new definitions I gave you? :)

**Comment by sthalik** [comment]
@MathijsG it appears there haven't been a single modification to the file in more than a year now. It's strange to say the least.

**Comment by Len62** [comment]
@sthalik Did you update the 04 version?

**Comment by sthalik** [comment]
@Len62 there's a small npclient fix in 04 but if it was working for you already, you don't need to update to 04. Actually I see why it might behave badly with both npclient and freetrack enabled.

@Len62, @MathijsG,

I've made a major refactoring for PT in order to support different camera drivers. It's gonna support the Wiimote via another developer's effort. It can also enable the free (as in beer) PS3 Eye driver as opposed to the current greedy scummy one. But mostly the Wiimote driver.

Don't get me wrong, PT worked almost immediately even after reorganizing most of the code. But I'd need a look by another party before releasing the new proper preview.

It works for me, including the preview, changing resolution during tracking, point detection and all else we want. But these heavy changes made me hesitant a bit.

The build filename is `opentrack-test-20180114_00-DO-NOT-USE.7z`, on http://opentrack.misaki.pl.

**Comment by Len62** [comment]
@sthalik Quick question. When you switch the output interface with tracking off, does it require a re-boot of OT to use the new selection? With this build I don't have any issues with the trackir interface, the freetrack interface does not work at all, unless I change to it from the trackir interface with OT already running. But, given my question, I'm wondering if it is actually using the freetrack one?

**Comment by MathijsG** [comment]
@sthalik I stumbled upon a steam community topic which mentioned that the original TrackIR also incorrectly recognizes Assetto Corsa as LockOn, so it that clarifies why the file isn't changed recently.

What free drivers do you mean for the PS3? I only know the $ 2 dollar ones. Can you get the full 75 hz with the default drivers? Sounds great though, maybe we should all intensively game with it for a few weeks with this new test build.

**Comment by Len62** [comment]
@MathijsG Go back up to the thread entries around 28 Sep and you'll see some info for the other driver.

**Comment by albertolg** [comment]
@sthalik  , what drivers are you refering? the libusb drivers or the others you linked (that are always the cleyedrivers)? I tested with the libusb but the camera is not recognized.

**Comment by sthalik** [comment]
@albertolg @MathijsG This one. You'd need set the driver for the ps3 eye device to libusb or libusbK: <<https://github.com/inspirit/PS3EYEDriver>>

@Len62 I'll look into it with Arma 3.

**Comment by albertolg** [comment]
@sthalik I installed the libusb with zadig, the camera works with the ps3eye_sdl.exe, but not with opentrack, is it working with your latest build opentrack-test-20180114_00-DO-NOT-USE.7z ?

thanks

**Comment by sthalik** [comment]
@Len62 @MathijsG can you check the "DO NOT USE" build if it works?

**Comment by sanpats** [comment]
I tried DO NOT USE build and it works fine. I only tested with old dlls in Arma 3 however. I tested new dlls only in OT.

**Comment by sthalik** [comment]
@albertolg no, it's not visible as a DirectShow camera. But we can interface with it.

**Comment by Len62** [comment]
@sthalik  The DO NOT USE build is working fine for me with the exception of the freetrack interface. Camera settings dialog works great, adjusting settings on the fly showed no ill effects.

**Comment by sthalik** [comment]
How much demand is there for the open PS3 Eye driver?

**Comment by albertolg** [comment]
I would like to get rid of the cleyedrivers, I use my ps3 eyecam only for opentrack

**Comment by Len62** [comment]
Agree with albertolg, but for people using their PS3 eye's for other uses, I guess the cleye driver would still be needed.

**Comment by sthalik** [comment]
Writing a DirectShow driver is annoying, as it's unnecessarily too complex because of Microsoft. But the current driver is used as a library so it's easier. But it's for the future. We could replace the paid driver completely.

Also, the Linux driver (this open Windows version) has better image quality than the Windows driver, no matter the resolution and Hz. It also has actually well-working white balance.

There are few improvements made by developers so it'd be nice to merge them into one project, and make my own. The paid PS3 Eye driver uses too much CPU as seen by the total consumption during tracking.

We'd also need opentrack infrastructure for managing non-native drivers. It's a bit of work but worthwhile.

**Comment by albertolg** [comment]
less CPU is always better... please include the DirectShow drivers, thanks

**Comment by sthalik** [comment]
It's not gonna be THAT significant. I doubt it could exceed 25%. Consider other pros/cons.

**Comment by sthalik** [comment]
Check [today's build](http://opentrack.misaki.pl). You have a third option for relative translation.

WW2 sim players, opentrack will temporarily disable reltrans while you aim (mapped yaw below 20, pitch below 20, roll below 7). I can increase the values if it's too little. Please try playing with reltrans on the "enabled when not aiming" choice and see if it works.

**Comment by albertolg** [comment]
but we always need to hope that with every windows update the cleyedrivers will still works

**Comment by sthalik** [comment]
Did you ever have problems? It worked for me.

**Comment by albertolg** [comment]
Nope, till now always worked, but I don't know if the guy will update his drivers if will noto work one day

**Comment by MathijsG** [comment]
I bought them for the $ 2 dollar, I'm not fundamentally against closed source commercial software ;-). I do always provide them to friends when I make them a headtracking device, although that's certainly not legal. It's always a good thing when OpenTrack becomes less dependent on third party payware, so I won't stop you from developing this driver :-), however you should always consider the effort when doing things. For many people 2 dollar wouldn't be a real showstopper, and they are also available on sites we all know already. But less CPU sounds interesting though because the headtracking already also needs a certain amount of CPU.

Will the microphone also work in your implementation? I like the ability of having it as a microphone, like I do have now, because the microphone is very high quality in comparing to other cheap devices.

**Comment by sthalik** [comment]
I don't know about the microphone... Question is whether the driver attaches to the composite device (bad) or just the inner camera (good).

It's gonna work for a decade or more since Microsoft keeps the old APIs around. It's also the same API other camera drivers use to expose them to Windows programs.

Anyway so many people depend on the camera (including this project) that a new solution will arise within weeks of a new Windows version breaking the webcam. Also remember that DLL wrappers and binary edits can fix drivers/DirectShow filters.

**Comment by sthalik** [comment]
Check out the new build. @cpuwolf wrote us a Wiimote tracker.

**Comment by Len62** [comment]
Interesting addition. Will have to try and find an old Wiimote, there is a Wii system stashed away somewhere in my house.

**Comment by sthalik** [comment]
Please give testing to the `opentrack-test-20180216_00-DO-NOT-USE` build. I've finally done a cleanup to the pipeline code. It's what drives the tracker, centers, filters, and otherwise transforms the data, finally sending it to the protocol.

I already gave it some testing but it's a big problem to get everything right since it affects:

- tracker data
- relative translation
- dynamic neck
- mapping from raw data
- filtering
- centering on keypress and startup
- "toggle" keybinding (should be renamed "hold" eventually)
- "zero" keybinding

Due to interaction of all the options it's a great pain to get it right. However the code was in such a bad shape it had to be done, or has to be postponed into the future.

Please give the build a try and see whether any of the options you use are broken, or behaving differently.

It's the same build as the other one today. The other build has no pipeline changes.

**Comment by Len62** [comment]
@sthalik Gave the do not use version a try. Maybe some issues w/point extraction/colors, will do some more testing. Biggest issue is tracking is not working in either IL2-BOS, or DCSW. Tracker is working, shortcut keys are working, just not working in game.

**Comment by albertolg** [comment]
confirmed not working

**Comment by sanpats** [comment]
I have no problem (Do not use version) with Arma 3, but I used old NPClient.dll and NPClient64.dll to avoid Arma 3 anti-cheat. I suppose it is not relate to the pipeline code that needed testing.

I will test it with Falcon BMS later tonight.

**Comment by sthalik** [comment]
It's not related to pipeline code. I'll check both of these titles.

**Comment by sanpats** [comment]
Seems to work fine for BMS as well, for me at least.

**Comment by sthalik** [comment]
@Len62 backing off the experimental point extraction change.

@albertolg @Len62 there was a bug indeed. See the new version with `_11` at the end. I'm deleting the old one. 

The new build doesn't have `DO-NOT-USE` in the name. New pipeline code may have unknown minor bugs but it works in general. Please report any issues!

**Comment by sthalik** [comment]
Please see #749 for `NPClient.dll` fix testing that was long coming.

**Comment by Len62** [comment]
No issues with the 11 version. Will test out the #749 issue. Thanks for the quick fix.

**Comment by albertolg** [comment]
yes, I also tested the new build with DCS and is working fine

**Comment by sthalik** [comment]
I'm releasing 2.3.10 without the following changes:

- pipeline rework
- NPClient work

This is because it was too long after the last release.

**Comment by sthalik** [comment]
I'm releasing 2.3.10 without the following changes:

- pipeline rework
- NPClient work

This is because it was too long after the last release.

**Comment by sthalik** [comment]
Please check the **2.3.10** build on http://opentrack.misaki.pl for any last-minute bugs. Thanks!

**Comment by sthalik** [comment]
Just to be clear, I backed out the NPClient and pipeline changes for hopefully the last 2.3 release.

**Comment by Len62** [comment]
I've been running the 2.3.10 tryit since it became available. I haven't noticed any ill effects with my set-up.

**Comment by sanpats** [comment]
2.3.10 works well with ArmA 3 & Falcon BMS.

**Comment by sthalik** [comment]
Version `2.3.10` has been released. Unlike the `2.4.0` effort, experimental changes are reverted. The `NPClient` bug is still here, the pipeline's not reworked.

**Comment by sthalik** [comment]
Please see #753, @Len62 @MathijsG. I need someone to complain so that I actually finish it. All that remains is connecting it to the tracking process.

**Comment by sniperwolfpk5** [comment]
I have tested it in DCS World working like a charm. Really appreciate your hard work. It is much better than 2.3.9.

**Comment by sthalik** [comment]
Please see https://github.com/opentrack/opentrack/issues/759#issue-303741090.

**Comment by sthalik** [comment]
Bad news, I gotta release the next build due to a FSX bug. Let it be current `unstable`, sans NPClient changes.

**Comment by Tonymazzz** [comment]
Guys i need control pc mouse with the gyro if my  phone, tried lots of config and no one works straight, its like doing curves... Help me please
